File: ts/S0003.ts
function lengthOfLongestSubstring(s: string): number {
    let str = '', result = 0;
    for (let i = 0; i < s.length; i++) {
        const index = str.indexOf(s[i]);
        if (index === -1) {
            str += s[i];
            result = Math.max(result, str.length);
        } else {
            str = str.slice(index + 1);
            str += s[i];
        }
    }
    return result;
};

console.log(lengthOfLongestSubstring("abcabc"));

File: go/S0003.go
package main

import "fmt"

func indexOf(s string, key byte) int {
	for i := 0; i < len(s); i++ {
		if s[i] == string(key) {
			return i
		}
	}
	return -1
}

func lengthOfLongestSubstring(s string) int {
	str := ""
	result := 0
	for i := 0; i < len(s); i++ {
		index := indexOf(str, s[i])
		if index == -1 {
			str += string(s[i])
			if result < len(str) {
				result = len(str)
			} else {
				str = str[index+1:]
				str += string(s[i])
			}
		}
	}
	return result
}

func main() {
	fmt.Println(lengthOfLongestSubstring("abcabdbb"))
}
